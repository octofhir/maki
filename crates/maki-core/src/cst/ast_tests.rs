//! Tests for typed AST layer

#[cfg(test)]
mod tests {
    use crate::cst::{
        ast::{AstNode, Document, Rule},
        parse_fsh,
    };

    #[test]
    fn test_profile_basic() {
        let source = r#"Profile: MyPatient
Parent: Patient
Id: my-patient
Title: "My Patient Profile"
Description: "A custom patient profile""#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");

        assert_eq!(profile.name().as_deref(), Some("MyPatient"));
        assert_eq!(
            profile.parent().unwrap().value().as_deref(),
            Some("Patient")
        );
        assert_eq!(profile.id().unwrap().value().as_deref(), Some("my-patient"));
        assert_eq!(
            profile.title().unwrap().value().as_deref(),
            Some("My Patient Profile")
        );
        assert_eq!(
            profile.description().unwrap().value().as_deref(),
            Some("A custom patient profile")
        );
    }

    #[test]
    #[ignore] // TODO: Rule path() API needs fixing - parser doesn't create Path nodes properly
    fn test_profile_with_rules() {
        let source = r#"Profile: MyPatient
Parent: Patient

* identifier 1..* MS
* name 1..1 MS SU
* gender from GenderValueSet (required)"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");

        let rules: Vec<_> = profile.rules().collect();
        assert_eq!(rules.len(), 3, "Should have 3 rules");

        // First rule: cardinality
        match &rules[0] {
            Rule::Card(card) => {
                assert_eq!(card.path().unwrap().as_string(), "identifier");
                assert!(card.cardinality().unwrap().contains("1..*"));
                assert!(card.flags().contains(&"MS".to_string()));
            }
            _ => panic!("Expected CardRule"),
        }

        // Second rule: cardinality with multiple flags
        match &rules[1] {
            Rule::Card(card) => {
                assert_eq!(card.path().unwrap().as_string(), "name");
                assert!(card.cardinality().unwrap().contains("1..1"));
                let flags = card.flags();
                assert!(flags.contains(&"MS".to_string()));
                assert!(flags.contains(&"SU".to_string()));
            }
            _ => panic!("Expected CardRule"),
        }

        // Third rule: valueset binding
        match &rules[2] {
            Rule::ValueSet(vs) => {
                assert_eq!(vs.path().unwrap().as_string(), "gender");
                assert_eq!(vs.value_set().as_deref(), Some("GenderValueSet"));
                assert_eq!(vs.strength().as_deref(), Some("required"));
            }
            _ => panic!("Expected ValueSetRule"),
        }
    }

    #[test]
    fn test_extension_basic() {
        let source = r#"Extension: MyExtension
Id: my-extension
Title: "My Extension"
Description: "A custom extension""#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let extension = doc.extensions().next().expect("Should have an extension");

        assert_eq!(extension.name().as_deref(), Some("MyExtension"));
        assert_eq!(
            extension.id().unwrap().value().as_deref(),
            Some("my-extension")
        );
        assert_eq!(
            extension.title().unwrap().value().as_deref(),
            Some("My Extension")
        );
        assert_eq!(
            extension.description().unwrap().value().as_deref(),
            Some("A custom extension")
        );
    }

    #[test]
    fn test_valueset_basic() {
        let source = r#"ValueSet: MyValueSet
Id: my-valueset
Title: "My Value Set"
Description: "A custom value set""#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let valueset = doc.value_sets().next().expect("Should have a valueset");

        assert_eq!(valueset.name().as_deref(), Some("MyValueSet"));

        let id_clause = valueset.id();
        if id_clause.is_none() {
            eprintln!("ValueSet CST: {:#?}", valueset.syntax());
        }

        assert_eq!(id_clause.unwrap().value().as_deref(), Some("my-valueset"));
        assert_eq!(
            valueset.title().unwrap().value().as_deref(),
            Some("My Value Set")
        );
        assert_eq!(
            valueset.description().unwrap().value().as_deref(),
            Some("A custom value set")
        );
    }

    #[test]
    fn test_codesystem_basic() {
        let source = r#"CodeSystem: MyCodeSystem
Id: my-codesystem
Title: "My Code System"
Description: "A custom code system""#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let codesystem = doc.code_systems().next().expect("Should have a codesystem");

        assert_eq!(codesystem.name().as_deref(), Some("MyCodeSystem"));
        assert_eq!(
            codesystem.id().unwrap().value().as_deref(),
            Some("my-codesystem")
        );
        assert_eq!(
            codesystem.title().unwrap().value().as_deref(),
            Some("My Code System")
        );
        assert_eq!(
            codesystem.description().unwrap().value().as_deref(),
            Some("A custom code system")
        );
    }

    #[test]
    fn test_alias_basic() {
        let source = r#"Alias: SCT = http://snomed.info/sct
Alias: LOINC = http://loinc.org"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let aliases: Vec<_> = doc.aliases().collect();

        assert_eq!(aliases.len(), 2);

        assert_eq!(aliases[0].name().as_deref(), Some("SCT"));
        assert_eq!(
            aliases[0].value().as_deref(),
            Some("http://snomed.info/sct")
        );

        assert_eq!(aliases[1].name().as_deref(), Some("LOINC"));
        assert_eq!(aliases[1].value().as_deref(), Some("http://loinc.org"));
    }

    #[test]
    fn test_multiple_profiles() {
        let source = r#"Profile: FirstProfile
Parent: Patient

Profile: SecondProfile
Parent: Observation"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profiles: Vec<_> = doc.profiles().collect();

        assert_eq!(profiles.len(), 2);
        assert_eq!(profiles[0].name().as_deref(), Some("FirstProfile"));
        assert_eq!(
            profiles[0].parent().unwrap().value().as_deref(),
            Some("Patient")
        );

        assert_eq!(profiles[1].name().as_deref(), Some("SecondProfile"));
        assert_eq!(
            profiles[1].parent().unwrap().value().as_deref(),
            Some("Observation")
        );
    }

    #[test]
    #[ignore] // TODO: Path parsing needs implementation
    fn test_path_segments() {
        let source = r#"Profile: MyProfile
Parent: Patient

* name.given 1..1"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");
        let rule = profile.rules().next().expect("Should have a rule");

        match rule {
            Rule::Card(card) => {
                let path = card.path().expect("Should have path");
                assert_eq!(path.as_string(), "name.given");
                let segments = path.segments();
                assert_eq!(segments, vec!["name", "given"]);
            }
            _ => panic!("Expected CardRule"),
        }
    }

    #[test]
    #[ignore] // TODO: Fixed value rule parsing needs implementation
    fn test_fixed_value_rule() {
        let source = r#"Profile: MyProfile
Parent: Patient

* active = true
* gender = "male""#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");
        let rules: Vec<_> = profile.rules().collect();

        assert_eq!(rules.len(), 2);

        // First fixed value: boolean
        match &rules[0] {
            Rule::FixedValue(fv) => {
                assert_eq!(fv.path().unwrap().as_string(), "active");
                assert_eq!(fv.value().as_deref(), Some("true"));
            }
            _ => panic!("Expected FixedValueRule"),
        }

        // Second fixed value: string
        match &rules[1] {
            Rule::FixedValue(fv) => {
                assert_eq!(fv.path().unwrap().as_string(), "gender");
                assert_eq!(fv.value().as_deref(), Some("male"));
            }
            _ => panic!("Expected FixedValueRule"),
        }
    }

    #[test]
    fn test_real_patient_profile() {
        // Use a real example from golden tests
        let source = include_str!("../../../../examples/patient-profile.fsh");
        let (cst, errors) = parse_fsh(source);

        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");

        // Verify we can extract metadata
        assert!(profile.name().is_some());
        assert!(profile.parent().is_some());
        assert!(profile.id().is_some());

        // Verify we can iterate rules
        let rule_count = profile.rules().count();
        assert!(rule_count > 0, "Should have at least one rule");
    }

    #[test]
    fn test_profile_without_optional_metadata() {
        let source = r#"Profile: MinimalProfile
Parent: Patient"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        let doc = Document::cast(cst).expect("Should be a document");
        let profile = doc.profiles().next().expect("Should have a profile");

        assert_eq!(profile.name().as_deref(), Some("MinimalProfile"));
        assert_eq!(
            profile.parent().unwrap().value().as_deref(),
            Some("Patient")
        );
        assert!(profile.id().is_none(), "Id should be None");
        assert!(profile.title().is_none(), "Title should be None");
        assert!(
            profile.description().is_none(),
            "Description should be None"
        );
    }

    #[test]
    fn test_ast_preserves_cst_lossless() {
        let source = r#"Profile:  MyPatient   // Extra spaces!
Parent: Patient
Id: my-patient

* name 1..1 MS"#;

        let (cst, errors) = parse_fsh(source);
        assert!(errors.is_empty());

        // Create typed AST
        let doc = Document::cast(cst.clone()).expect("Should be a document");
        let _profile = doc.profiles().next().expect("Should have a profile");

        // Verify CST is still lossless
        assert_eq!(
            cst.text().to_string(),
            source,
            "CST should still be lossless"
        );
    }
}
