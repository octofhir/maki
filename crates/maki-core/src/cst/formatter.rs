//! CST-based formatter for FHIR Shorthand
//!
//! This module provides formatting capabilities using the typed AST layer
//! over the lossless CST. It can:
//! - Normalize whitespace while preserving comments
//! - Align carets in rules
//! - Format metadata clauses consistently
//! - Maintain idempotency (format(format(x)) == format(x))
//!
//! # Example
//!
//! ```rust,ignore
//! use maki_core::cst::{parse_fsh, formatter::format_document};
//!
//! let source = "Profile:MyPatient\nParent:Patient\n*name 1..1";
//! let formatted = format_document(source, &FormatOptions::default());
//!
//! assert_eq!(formatted, r#"Profile: MyPatient
//! Parent: Patient
//!
//! * name 1..1
//! "#);
//! ```

use super::{
    ast::{Alias, AstNode, CodeSystem, Document, Extension, Profile, Rule, ValueSet},
    parse_fsh,
};

/// Formatting options
#[derive(Debug, Clone)]
pub struct FormatOptions {
    /// Number of spaces for indentation
    pub indent_size: usize,

    /// Whether to align carets in rules
    pub align_carets: bool,

    /// Maximum line length before wrapping (0 = no limit)
    pub max_line_length: usize,

    /// Whether to add blank line before rules
    pub blank_line_before_rules: bool,

    /// Whether to preserve existing blank lines
    pub preserve_blank_lines: bool,
}

impl Default for FormatOptions {
    fn default() -> Self {
        Self {
            indent_size: 2,
            align_carets: true,
            max_line_length: 100,
            blank_line_before_rules: true,
            preserve_blank_lines: false,
        }
    }
}

/// Format context tracks state during formatting
struct FormatContext {
    options: FormatOptions,
    output: String,
    indent_level: usize,
}

impl FormatContext {
    fn new(options: FormatOptions) -> Self {
        Self {
            options,
            output: String::new(),
            indent_level: 0,
        }
    }

    /// Write a line with current indentation
    fn writeln(&mut self, text: &str) {
        if !text.is_empty() {
            self.write_indent();
            self.output.push_str(text);
        }
        self.output.push('\n');
    }

    /// Write current indentation
    fn write_indent(&mut self) {
        let indent = " ".repeat(self.indent_level * self.options.indent_size);
        self.output.push_str(&indent);
    }

    /// Add a blank line
    fn blank_line(&mut self) {
        self.output.push('\n');
    }

    /// Get the final formatted output
    fn finish(self) -> String {
        self.output
    }
}

/// Format a complete FSH document
pub fn format_document(source: &str, options: &FormatOptions) -> String {
    let (cst, errors) = parse_fsh(source);

    if !errors.is_empty() {
        // If there are parse errors, return original source
        // In a real implementation, we might want to do partial formatting
        return source.to_string();
    }

    let doc = match Document::cast(cst) {
        Some(doc) => doc,
        None => return source.to_string(),
    };

    let mut ctx = FormatContext::new(options.clone());

    // Format aliases first (they typically come first)
    let mut first = true;
    for alias in doc.aliases() {
        if !first {
            ctx.blank_line();
        }
        format_alias(&mut ctx, &alias);
        first = false;
    }

    // Add blank line between aliases and definitions
    if !first {
        ctx.blank_line();
    }

    // Format profiles
    first = true;
    for profile in doc.profiles() {
        if !first {
            ctx.blank_line();
        }
        format_profile(&mut ctx, &profile);
        first = false;
    }

    // Format extensions
    for extension in doc.extensions() {
        if !first {
            ctx.blank_line();
        }
        format_extension(&mut ctx, &extension);
        first = false;
    }

    // Format value sets
    for valueset in doc.value_sets() {
        if !first {
            ctx.blank_line();
        }
        format_valueset(&mut ctx, &valueset);
        first = false;
    }

    // Format code systems
    for codesystem in doc.code_systems() {
        if !first {
            ctx.blank_line();
        }
        format_codesystem(&mut ctx, &codesystem);
        first = false;
    }

    ctx.finish()
}

/// Format an alias declaration
fn format_alias(ctx: &mut FormatContext, alias: &Alias) {
    let name = alias.name().unwrap_or_default();
    let value = alias.value().unwrap_or_default();
    ctx.writeln(&format!("Alias: {name} = {value}"));
}

/// Format a profile definition
fn format_profile(ctx: &mut FormatContext, profile: &Profile) {
    // Profile header
    let name = profile.name().unwrap_or_default();
    ctx.writeln(&format!("Profile: {name}"));

    // Metadata clauses
    if let Some(parent) = profile.parent() {
        let value = parent.value().unwrap_or_default();
        ctx.writeln(&format!("Parent: {value}"));
    }

    if let Some(id) = profile.id() {
        let value = id.value().unwrap_or_default();
        ctx.writeln(&format!("Id: {value}"));
    }

    if let Some(title) = profile.title() {
        let value = title.value().unwrap_or_default();
        ctx.writeln(&format!("Title: \"{value}\""));
    }

    if let Some(description) = profile.description() {
        let value = description.value().unwrap_or_default();
        ctx.writeln(&format!("Description: \"{value}\""));
    }

    // Rules section
    let rules: Vec<_> = profile.rules().collect();
    if !rules.is_empty() {
        if ctx.options.blank_line_before_rules {
            ctx.blank_line();
        }

        // Calculate caret alignment if needed
        let caret_column = if ctx.options.align_carets {
            calculate_caret_alignment(&rules)
        } else {
            0
        };

        for rule in rules {
            format_rule(ctx, &rule, caret_column);
        }
    }
}

/// Format an extension definition
fn format_extension(ctx: &mut FormatContext, extension: &Extension) {
    let name = extension.name().unwrap_or_default();
    ctx.writeln(&format!("Extension: {name}"));

    if let Some(id) = extension.id() {
        let value = id.value().unwrap_or_default();
        ctx.writeln(&format!("Id: {value}"));
    }

    if let Some(title) = extension.title() {
        let value = title.value().unwrap_or_default();
        ctx.writeln(&format!("Title: \"{value}\""));
    }

    if let Some(description) = extension.description() {
        let value = description.value().unwrap_or_default();
        ctx.writeln(&format!("Description: \"{value}\""));
    }

    let rules: Vec<_> = extension.rules().collect();
    if !rules.is_empty() {
        if ctx.options.blank_line_before_rules {
            ctx.blank_line();
        }

        let caret_column = if ctx.options.align_carets {
            calculate_caret_alignment(&rules)
        } else {
            0
        };

        for rule in rules {
            format_rule(ctx, &rule, caret_column);
        }
    }
}

/// Format a value set definition
fn format_valueset(ctx: &mut FormatContext, valueset: &ValueSet) {
    let name = valueset.name().unwrap_or_default();
    ctx.writeln(&format!("ValueSet: {name}"));

    if let Some(id) = valueset.id() {
        let value = id.value().unwrap_or_default();
        ctx.writeln(&format!("Id: {value}"));
    }

    if let Some(title) = valueset.title() {
        let value = title.value().unwrap_or_default();
        ctx.writeln(&format!("Title: \"{value}\""));
    }

    if let Some(description) = valueset.description() {
        let value = description.value().unwrap_or_default();
        ctx.writeln(&format!("Description: \"{value}\""));
    }
}

/// Format a code system definition
fn format_codesystem(ctx: &mut FormatContext, codesystem: &CodeSystem) {
    let name = codesystem.name().unwrap_or_default();
    ctx.writeln(&format!("CodeSystem: {name}"));

    if let Some(id) = codesystem.id() {
        let value = id.value().unwrap_or_default();
        ctx.writeln(&format!("Id: {value}"));
    }

    if let Some(title) = codesystem.title() {
        let value = title.value().unwrap_or_default();
        ctx.writeln(&format!("Title: \"{value}\""));
    }

    if let Some(description) = codesystem.description() {
        let value = description.value().unwrap_or_default();
        ctx.writeln(&format!("Description: \"{value}\""));
    }
}

/// Calculate the column position for caret alignment
fn calculate_caret_alignment(rules: &[Rule]) -> usize {
    // Find the longest path in all rules
    let max_path_len = rules
        .iter()
        .filter_map(|rule| {
            let path = match rule {
                Rule::Card(r) => r.path().map(|p| p.as_string()),
                Rule::Flag(r) => r.path().map(|p| p.as_string()),
                Rule::ValueSet(r) => r.path().map(|p| p.as_string()),
                Rule::FixedValue(r) => r.path().map(|p| p.as_string()),
                Rule::Path(r) => r.path().map(|p| p.as_string()),
                Rule::Contains(r) => r.path().map(|p| p.as_string()),
                Rule::Only(r) => r.path().map(|p| p.as_string()),
                Rule::Obeys(r) => r.path().map(|p| p.as_string()),
                Rule::AddElement(r) => r.path().map(|p| p.as_string()),
                Rule::Mapping(r) => r.path().map(|p| p.as_string()),
            };
            path.map(|p| p.len())
        })
        .max()
        .unwrap_or(0);

    // Caret column = "* " (2) + max_path_len + " " (1)
    2 + max_path_len + 1
}

/// Format a single rule
fn format_rule(ctx: &mut FormatContext, rule: &Rule, caret_column: usize) {
    match rule {
        Rule::Card(card) => {
            let path = card.path().map(|p| p.as_string()).unwrap_or_default();
            let cardinality = card.cardinality().unwrap_or_default();
            let flags = card.flags();

            if caret_column > 0 {
                // Aligned format: "* path      1..1 MS"
                let padding = caret_column.saturating_sub(2 + path.len());
                let rule_text = format!(
                    "* {}{}{}{}",
                    path,
                    " ".repeat(padding),
                    cardinality,
                    if flags.is_empty() {
                        String::new()
                    } else {
                        format!(" {}", flags.join(" "))
                    }
                );
                ctx.writeln(&rule_text);
            } else {
                // Non-aligned format: "* path 1..1 MS"
                let rule_text = format!(
                    "* {} {}{}",
                    path,
                    cardinality,
                    if flags.is_empty() {
                        String::new()
                    } else {
                        format!(" {}", flags.join(" "))
                    }
                );
                ctx.writeln(&rule_text);
            }
        }

        Rule::Flag(flag) => {
            let path = flag.path().map(|p| p.as_string()).unwrap_or_default();
            let flags = flag.flags();

            if caret_column > 0 {
                let padding = caret_column.saturating_sub(2 + path.len());
                ctx.writeln(&format!(
                    "* {}{}{}",
                    path,
                    " ".repeat(padding),
                    flags.join(" ")
                ));
            } else {
                ctx.writeln(&format!("* {} {}", path, flags.join(" ")));
            }
        }

        Rule::ValueSet(vs) => {
            let path = vs.path().map(|p| p.as_string()).unwrap_or_default();
            let valueset = vs.value_set().unwrap_or_default();
            let strength = vs.strength();

            let rule_text = if let Some(s) = strength {
                format!("* {path} from {valueset} ({s})")
            } else {
                format!("* {path} from {valueset}")
            };
            ctx.writeln(&rule_text);
        }

        Rule::FixedValue(fv) => {
            let path = fv.path().map(|p| p.as_string()).unwrap_or_default();
            let value = fv.value().unwrap_or_default();

            // Check if value needs quotes (not a number or boolean)
            let formatted_value =
                if value.parse::<i64>().is_ok() || value == "true" || value == "false" {
                    value
                } else {
                    format!("\"{value}\"")
                };

            ctx.writeln(&format!("* {path} = {formatted_value}"));
        }

        Rule::Path(path_rule) => {
            let path = path_rule.path().map(|p| p.as_string()).unwrap_or_default();
            ctx.writeln(&format!("* {path}"));
        }

        Rule::Contains(contains_rule) => {
            let path = contains_rule
                .path()
                .map(|p| p.as_string())
                .unwrap_or_default();
            let items = contains_rule.items();
            ctx.writeln(&format!("* {} contains {}", path, items.join(" and ")));
        }

        Rule::Only(only_rule) => {
            let path = only_rule.path().map(|p| p.as_string()).unwrap_or_default();
            let types = only_rule.types();
            ctx.writeln(&format!("* {} only {}", path, types.join(" or ")));
        }

        Rule::Obeys(obeys_rule) => {
            let path = obeys_rule.path().map(|p| p.as_string()).unwrap_or_default();
            let invariants = obeys_rule.invariants();
            ctx.writeln(&format!("* {} obeys {}", path, invariants.join(" and ")));
        }

        Rule::AddElement(add_rule) => {
            let path = add_rule.path().map(|p| p.as_string()).unwrap_or_default();
            let cardinality = add_rule.cardinality().unwrap_or_default();
            let flags = add_rule.flags();
            let types = add_rule.types();
            let short = add_rule.short();
            let definition = add_rule.definition();

            // Format: * path card flags type "short" "definition"
            let mut rule_text = format!("* {} {}", path, cardinality);

            if !flags.is_empty() {
                rule_text.push(' ');
                rule_text.push_str(&flags.join(" "));
            }

            if !types.is_empty() {
                rule_text.push(' ');
                rule_text.push_str(&types.join(" or "));
            }

            if let Some(short_desc) = short {
                rule_text.push_str(&format!(" \"{}\"", short_desc));
            }

            if let Some(def_desc) = definition {
                rule_text.push_str(&format!(" \"{}\"", def_desc));
            }

            ctx.writeln(&rule_text);
        }

        Rule::Mapping(mapping_rule) => {
            let path = mapping_rule
                .path()
                .map(|p| p.as_string())
                .unwrap_or_default();
            let map = mapping_rule.map().unwrap_or_default();
            let comment = mapping_rule.comment();
            let language = mapping_rule.language();

            // Format: * path -> "target" "comment" #language
            let mut rule_text = format!("* {} -> \"{}\"", path, map);

            if let Some(comment_text) = comment {
                rule_text.push_str(&format!(" \"{}\"", comment_text));
            }

            if let Some(lang) = language {
                rule_text.push_str(&format!(" #{}", lang));
            }

            ctx.writeln(&rule_text);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_basic_profile() {
        let source = "Profile:MyPatient\nParent:Patient\nId:my-patient";
        let formatted = format_document(source, &FormatOptions::default());

        assert_eq!(
            formatted,
            "Profile: MyPatient\nParent: Patient\nId: my-patient\n"
        );
    }

    #[test]
    #[ignore] // TODO: Formatter doesn't output rule paths - needs fixing
    fn test_format_profile_with_rules() {
        let source = "Profile:MyPatient\nParent:Patient\n*name 1..1 MS\n*gender 1..1";
        let formatted = format_document(source, &FormatOptions::default());

        let expected = r#"Profile: MyPatient
Parent: Patient

* name   1..1 MS
* gender 1..1
"#;

        assert_eq!(formatted, expected);
    }

    #[test]
    fn test_format_idempotency() {
        let source = r#"Profile: MyPatient
Parent: Patient

* name 1..1 MS
* gender 1..1
"#;

        let formatted1 = format_document(source, &FormatOptions::default());
        let formatted2 = format_document(&formatted1, &FormatOptions::default());

        assert_eq!(formatted1, formatted2, "Formatting should be idempotent");
    }

    #[test]
    fn test_format_alias() {
        let source = "Alias:SCT=http://snomed.info/sct";
        let formatted = format_document(source, &FormatOptions::default());

        assert_eq!(formatted, "Alias: SCT = http://snomed.info/sct\n\n");
    }

    #[test]
    #[ignore] // TODO: Formatter doesn't output rule paths - needs fixing
    fn test_caret_alignment() {
        let source = r#"Profile: Test
Parent: Patient

* identifier 1..*
* name 1..1
* birthDate 0..1
"#;

        let formatted = format_document(source, &FormatOptions::default());

        // All carets should be aligned
        let lines: Vec<&str> = formatted.lines().collect();
        assert!(lines[3].starts_with("* identifier "));
        assert!(lines[4].starts_with("* name       "));
        assert!(lines[5].starts_with("* birthDate  "));
    }

    #[test]
    #[ignore] // TODO: Formatter doesn't output rule paths - needs fixing
    fn test_no_caret_alignment() {
        let source = "Profile:Test\nParent:Patient\n*identifier 1..*\n*name 1..1";

        let options = FormatOptions {
            align_carets: false,
            ..Default::default()
        };

        let formatted = format_document(source, &options);

        let expected = r#"Profile: Test
Parent: Patient

* identifier 1..*
* name 1..1
"#;

        assert_eq!(formatted, expected);
    }
}
