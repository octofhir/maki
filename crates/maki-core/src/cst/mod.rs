//! Concrete Syntax Tree (CST) for FHIR Shorthand
//!
//! This module implements a lossless syntax tree using the Rowan library.
//! The CST preserves all source information including whitespace, comments,
//! and formatting, enabling:
//! - Accurate source-to-source transformations (formatter)
//! - Precise autofixes that preserve formatting
//! - Better error recovery and diagnostics
//!
//! ## Architecture
//!
//! The CST uses Rowan's green/red tree pattern:
//!
//! - **Green Tree**: Immutable, position-independent storage
//!   - Stores actual source text with trivia (whitespace, comments)
//!   - Deduplicates identical subtrees for memory efficiency
//!   - Cheap to clone (uses Arc internally)
//!
//! - **Red Tree**: Dynamically constructed view with parent pointers
//!   - Created on-demand for traversal
//!   - Provides typed AST-like API
//!   - Enables efficient parent/sibling navigation
//!
//! ## Trivia Handling
//!
//! Trivia is attached to tokens:
//! - **Leading trivia**: Everything before a token (whitespace, comments)
//! - **Trailing trivia**: Everything until the next line break
//!
//! This enables lossless representation: `parse(source).text() == source`
//!
//! ## Example
//!
//! ```rust,ignore
//! use maki_core::cst::{FshSyntaxNode, FshSyntaxKind};
//!
//! // Parse FSH to CST
//! let (cst, errors) = parse_fsh("Profile: MyPatient // comment\nParent: Patient");
//!
//! // Verify lossless property
//! assert_eq!(cst.text().to_string(), "Profile: MyPatient // comment\nParent: Patient");
//!
//! // Navigate CST
//! for token in cst.descendants_with_tokens() {
//!     if let Some(comment) = token.as_token() {
//!         if comment.kind() == FshSyntaxKind::CommentLine {
//!             println!("Found comment: {}", comment.text());
//!         }
//!     }
//! }
//! ```

mod builder;
mod language;
mod lexer;
mod nodes;
mod parser;
mod syntax_kind;

pub mod ast;
pub mod formatter;

pub use builder::{CstBuilder, build_cst_from_tokens, parse_fsh_simple};
pub use formatter::{FormatOptions, format_document};
pub use language::FshLanguage;
pub use lexer::{CstLexResult, CstToken, lex_with_trivia};
pub use nodes::*;
pub use parser::parse_fsh;
pub use syntax_kind::FshSyntaxKind;

#[cfg(test)]
mod tests;

#[cfg(test)]
mod golden_tests;

#[cfg(test)]
mod ast_tests;

#[cfg(test)]
mod formatter_insta_tests;
#[cfg(test)]
mod formatter_snapshot_tests;
