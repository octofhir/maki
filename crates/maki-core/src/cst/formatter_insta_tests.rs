//! Snapshot tests for formatter using insta

#[cfg(test)]
mod tests {
    use crate::cst::formatter::{FormatOptions, format_document};
    use insta::assert_snapshot;

    #[test]
    fn test_format_patient_profile() {
        let source = include_str!("../../../../examples/patient-profile.fsh");
        let formatted = format_document(source, &FormatOptions::default());

        assert_snapshot!(formatted);
    }

    #[test]
    fn test_format_simple_profile() {
        let source = r#"Profile:MyPatient
Parent:Patient
Id:my-patient
Title:"My Patient Profile"

*identifier 1..* MS
*name 1..1 MS
*gender 1..1"#;

        let formatted = format_document(source, &FormatOptions::default());
        assert_snapshot!(formatted);
    }

    #[test]
    fn test_format_with_caret_rules() {
        let source = r#"Profile:Test
Parent:Patient

*^version = "1.0.0"
*^status = #active
*name 1..1"#;

        let formatted = format_document(source, &FormatOptions::default());
        assert_snapshot!(formatted);
    }

    #[test]
    fn test_format_extension_contains() {
        let source = r#"Profile:Test
Parent:Patient

*extension contains
    race 0..1 MS and
    ethnicity 0..1 MS"#;

        let formatted = format_document(source, &FormatOptions::default());
        assert_snapshot!(formatted);
    }

    #[test]
    fn test_format_valueset_binding() {
        let source = r#"Profile:Test
Parent:Patient

*gender from GenderValueSet (required)
*communication.language from AllLanguages (extensible)"#;

        let formatted = format_document(source, &FormatOptions::default());
        assert_snapshot!(formatted);
    }

    #[test]
    #[ignore] // TODO: Formatter doesn't output rule paths properly
    fn test_format_comprehensive_test() {
        let source = include_str!("../../../../examples/comprehensive-test.fsh");
        let formatted = format_document(source, &FormatOptions::default());

        assert_snapshot!(formatted);
    }

    #[test]
    fn test_format_idempotency_patient() {
        let source = include_str!("../../../../examples/patient-profile.fsh");
        let formatted1 = format_document(source, &FormatOptions::default());
        let formatted2 = format_document(&formatted1, &FormatOptions::default());

        assert_eq!(formatted1, formatted2, "Formatting should be idempotent");
    }

    #[test]
    fn test_format_no_alignment() {
        let source = r#"Profile:Test
Parent:Patient

*identifier 1..* MS
*name 1..1 MS
*gender 1..1"#;

        let options = FormatOptions {
            align_carets: false,
            ..Default::default()
        };

        let formatted = format_document(source, &options);
        assert_snapshot!(formatted);
    }
}
