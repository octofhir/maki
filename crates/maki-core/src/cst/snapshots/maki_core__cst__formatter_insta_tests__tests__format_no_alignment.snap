---
source: crates/maki-core/src/cst/formatter_insta_tests.rs
expression: formatted
---
Profile: Test
Parent: Patient

* identifier 1..* MS
* name 1..1 MS
* gender 1..1
